build: add `tests` extras & update `developer_docs.md`

alvarobartt opened the pull request:
# Description

Since the only option to setup a development environment was via `conda` and not all the users don't use `conda`, I've included the extra dependencies to run the tests as extras, to be able to easily install those via `pip` without forcing the users to rely on `conda`.

Besides that, I've also included some more information in the `developer_docs.md` to make it easier for the users to install everything and run the tests locally.

Additionally, while tackling this PR, some other minor things have been spotted and solved in other PRs:

* https://github.com/argilla-io/argilla/pull/3015
* https://github.com/argilla-io/argilla/pull/3012
* https://github.com/argilla-io/argilla/pull/3105

**Type of change**

- [X] Improvement (change adding some improvement to an existing functionality)
- [X] Documentation update

**How Has This Been Tested**

- [x] Tested in Python 3.8, 3.9 and 3.10 by installing everything as `pip install -e ".[server,postgresql,listeners,tests]"` and then running `pytest`
- [X] Tested that `conda` installation still works fine

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings

alvarobartt committed:
Install `psycopg2-binary` on macOS (darwin)

alvarobartt committed:
Standardize space in dependency-specifier versions

alvarobartt committed:
Format `pyproject.toml` file

alvarobartt committed:
Add `tests` extra dependencies

Tested in Python 3.8, 3.9 and 3.10 by installing everything as `pip install -e ".[server,postgresql,listeners,tests]` and then running `pytest`

alvarobartt committed:
Update `developer_docs.md` with `pip` install

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

gabrielmbmb commented on pyproject.toml:
```diff
@@ -6,8 +6,8 @@ build-backend = "setuptools.build_meta"
 readme = "README.md"
-requires-python = ">=3.7"
-license = {text = "Apache-2.0"}
+requires-python = ">=3.7"  # TODO: upgrade minimum Python version to 3.8, and pin maximum to < 3.12
```
We will deprecate `3.7` in the `1.9.0` release, right?

alvarobartt replied:
Yes I think so, didn't want to update that yet!

gabrielmbmb reviewed (approved):
Hey @alvarobartt, LGTM! This will be really helpful 😄 Maybe it would be good also to mention in the developer docs something about the dependencies needed to be installed to be able to build and install `psycopg2`.

alvarobartt committed:
Pin `flair` dependency to 0.12

alvarobartt committed:
Add `get_spans` to `FlairMonitor` due to upgrade

alvarobartt committed:
Put `flair` types under `typing.TYPE_CHECKING`

alvarobartt committed:
Fix `test_flair_monitoring` unit tests

alvarobartt committed:
Upgrade `flair` to 0.12 in `environment_dev.yml`

alvarobartt committed:
Revert `TYPE_CHECKING` usage in `FlairMonitor`

Breaks the unit tests because those imports previously included under `typing.TYPE_CHECKING` are imported from other module/s

alvarobartt committed:
Use `inspect.getfullargspec` for args & defaults

Use `inspect.getfullargspec` instead of the former approach which was more complex for no reason, since `getfullargspec` retrieves all the args and their defaults, if applicable

alvarobartt committed:
`shutil.rmtree` just if file exists in advance

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

alvarobartt committed:
Add missing return type-hints

alvarobartt committed:
Combine `_from_pretrained` & `from_pretrained` args

alvarobartt committed:
Update just existing `setfit` args in `update_config`

alvarobartt committed:
Apply suggestions from code review

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

alvarobartt committed:
Fix function-args retrieval and file-deletion after train (#3015)

# Description

Some `setfit` unit tests were not passing as it can be seen at
https://github.com/argilla-io/argilla/actions/runs/5113755403/jobs/9193313124,
with the exception `FileNotFound` which was due to the function defined
in `tests/training/helpers.py` named `train_with_cleanup` was not
properly cleaning the file, since it was trying to `shutil.rmtree` over
a file that could or could not exist. I assume that issue was not
appearing on other runs since probably the file was cached at some
point.

Then after solving that issue, another issue appeared `TypeError: got
multiple values for argument 'model_id'`, and `TypeError: got multiple
values for argument 'revision'`, which was also leading the `setfit`
tests to fail.

That was due to the function defined at `argilla/training/utils.py`
named `get_default_args` was not working as expected, since it was
having issues with the inheritance of the classmethod `from_pretrained`,
and adding some args that were not from the original function such as
`model_id` and `revision` which are args from the `_from_pretrained`
method, so on, since `from_pretrained` keeps the `kwargs` and then
propagates those to `_from_pretrained` (called as part of the return
statement), it was receiving the `model_id` and `revision` args set
during `from_pretrained`, as well as the ones propagated via `kwargs`
since those are not valid args in `from_pretrained`. So on, I've fixed
this by using `inspect.getfullargspec` instead, and that seems to work
without any issue when inheriting from multiple classes or even
inheriting from a MixIn.

**Type of change**

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

- [X] Re-run the `setfit` unit tests on Python 3.8, 3.9, and 3.10 and
all those are passing properly now

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] My changes generate no new warnings

alvarobartt committed:
chore: upgrade `flair` to 0.12 (#3012)

# Description

As older versions of `flair` (< 0.12) were having problems during the
installation, I've decided to upgrade `flair` to 0.12 and pin it.
Besides the package upgrade, some things had to be updated in the
`FlairMonitor` for the unit tests to pass when installing the latest
version of `flair`.

Some more information about the installation issue at
https://github.com/flairNLP/flair/issues/3129

**Type of change**

- [X] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

- [x] Updated the `FlairMonitoring` unit tests

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my
feature works

alvarobartt committed:
Merge branch 'develop' into feat/add-tests-extras

alvarobartt committed:
chore: upgrade `flair` to 0.12.2

0.12.0 solved the installation issues, but had a constraint over the `huggingface_hub` version to be 0.10.0 only, which was having some conflicts with `autotrain-advanced`, recently included

alvarobartt committed:
chore: upgrade `flair` to latest version (0.12.2) (#3105)

# Description

As of #3012, we decided to upgrade `flair` to 0.12.0 to avoid some known
installation issues with both dependencies: `tokenizers` and
`sentencepiece`. It was solved, but we recently included the
`ArgillaAutoTrainTrainer` at #2940 (kudos to @davidberenstein1957), and
it seems that `autotrain-advanced` dependency expects a greater
`huggingface_hub` version, while `flair` 0.12.0 had it pinned to
`==0.10.0` which was pretty restrictive and, so on, running into
conflicts with `autotrain-advanced`.

So on, it seems that `flair` 0.12.2 already defines `huggingface_hub`
dependency as `>=0.10.0` which is more relaxed, and doesn't run into any
conflict with the rest of the dependencies.

See the previous `huggingface_hub` dependency pinned to ==0.10.0 at
https://github.com/flairNLP/flair/blob/d13f8ea36e60c5ab1d7468893498db96d91fd4be/requirements.txt#L22,
and now being pinned to >=0.10.0 at
https://github.com/flairNLP/flair/blob/621e7dd58be8e3c5985176d9a7f54eb5c0cd8c88/requirements.txt#L22,
as of 0.12.2, since in 0.12.1 it's still pinned to ==0.10.0.

**Type of change**

- [X] Upgrade/Add dependency

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And
ideally, reference `tests`)

- [X] Installed via `pip` in Python 3.8, 3.9, 3.10, and 3.11; and with
`conda` in Python 3.8

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [ ] follows the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)
